MastodonSDK/Sources/MastodonUI/Helper/MediaLayoutHelper.swift

```diff
@@ -0,0 +1,314 @@
+			
+			// Two lines
+			for firstLine in 1...cnt-1 {
+				tries[[firstLine, cnt-firstLine]] = [
+					calculateMultiThumbsHeight(ratios: Array(ratiosCropped[..<firstLine]), width: maxWidth, margin: gap),
+					calculateMultiThumbsHeight(ratios: Array(ratiosCropped[firstLine...]), width: maxWidth, margin: gap)
+				]
+			}
+			
+			// Three lines
+			for firstLine in 1...cnt-2 {
+				for secondLine in 1...cnt-firstLine-1 {
+					tries[[firstLine, secondLine, cnt-firstLine-secondLine]] = [
+						calculateMultiThumbsHeight(ratios: Array(ratiosCropped[..<firstLine]), width: maxWidth, margin: gap),
+						calculateMultiThumbsHeight(ratios: Array(ratiosCropped[firstLine..<firstLine+secondLine]), width: maxWidth, margin: gap),
+						calculateMultiThumbsHeight(ratios: Array(ratiosCropped[(firstLine+secondLine)...]), width: maxWidth, margin: gap)
+					]
+				}
+			}
+			
+			let realMaxHeight = min(maxWidth, maxHeight)
+			
+			var optConf: [Int] = []
+			var optDiff: Float = Float.greatestFiniteMagnitude
+			
+			for (conf, heights) in tries {
+				let confH: Float = heights.reduce(gap*Float(heights.count-1), +)
+				var confDiff = abs(confH-realMaxHeight)
+				if conf.count>1 && (conf[0]>conf[1] || (conf.count>2 && conf[1]>conf[2])) {
+					confDiff *= 1.1
+				}
+				if confDiff<optDiff {
+					optConf = conf
+					optDiff = confDiff
+				}
+			}
+			
+			var thumbsRemain: [MastodonAttachment] = Array(attachments)
+			var ratiosRemain: [Float] = Array(ratiosCropped)
+			let optHeights = tries[optConf]!
+			var totalHeight: Float = 0.0
+			var rowSizes: [Int] = []
+			var gridLineOffsets: [Int] = []
+			var rowTiles: [[MediaLayoutResult.Tile]] = []
+			
+			for (i, lineChunksNum) in optConf.enumerated() {
+				var lineThumbs: [MastodonAttachment] = []
+				for _ in 0..<lineChunksNum {
+					lineThumbs.append(thumbsRemain.removeFirst())
+				}
+				let lineHeight = optHeights[i]
+				totalHeight += lineHeight
+				rowSizes.append(Int(lineHeight.rounded()))
+				var totalWidth: Int = 0
+				var row: [MediaLayoutResult.Tile] = []
+				for (j, _) in lineThumbs.enumerated() {
+					let thumbRatio = ratiosRemain.removeFirst()
+					let w: Float = j==lineThumbs.count-1 ? (maxWidth-Float(totalWidth)) : (thumbRatio*lineHeight)
+					totalWidth += Int(w.rounded())
+					if j<lineThumbs.count-1 && !gridLineOffsets.contains(totalWidth) {
+						gridLineOffsets.append(totalWidth)
+					}
+					var tile = MediaLayoutResult.Tile(colSpan: 1, rowSpan: 1, startCol: 0, startRow: i)
+					tile.width = Int(w.rounded())
+					row.append(tile)
+				}
+				rowTiles.append(row)
+			}
+			
+			gridLineOffsets = gridLineOffsets.sorted()
+			gridLineOffsets.append(Int(maxWidth))
+			
+			var columnSizes: [Int] = [gridLineOffsets[0]]
+			for (i, offset) in gridLineOffsets[1...].enumerated() {
+				columnSizes.append(offset - gridLineOffsets[i]) // i is already offset by one here
+			}
+			
+			for (row, _) in rowTiles.enumerated() {
```

Comment:
I don’t think you need the `enumerated` here. If you want the values:

```suggestion
			for row in rowTiles {
```
 or if you want the indices:

```suggestion
			for row in 0..<rowTiles.count {
```